os/include/tinyara/preference.h

```diff
@@ -70,21 +71,18 @@ enum preference_result_error_e {
 #define PREFERENCE_CBMQ_LEN         16
 #define PREFERNENCE_CBMSG_MAX       32
 
-struct value_data_s {
+struct value_attr_s {
```

Comment:
In my opinion something should be changed,
1. I think this struct is not necessary.
 You have to define APIs first, for example,
 - pref_create_subkey(char *subkey) // Create directory for key. c.f /mnt/prefkey/appA 
 - pref_open_key(char* key);
 - pref_get_key_value(char* key, char *buf, ssize_t len);
 ....
 Then user will use these apis without this struct.
 also for file system, only key & length is needed to get whole of key value.
 If there's any reason to keep any of structure, then provide handler which is hide from user.
2. Will you open whole of structure & definition to user? I think many things here, can be hided from user.

Reply:
1. I understand what you consider. I will move preference operations to driver to support them in kernel as you said. And then it would be implemented with general usages like buf, len.

2. I think those are needed because user can't call preference functions directly. If not, all functions should be registered to syscall or user needs to pass all parameters to functions respectively.

Reply:
1. Sorry, I was confused. I think driver is not appropriate for preference.
I will update preference functions for kernel and general usage as you said.